### ISSUE TITLE
Federating Models

### ISSUE THREAD

**Issue by LukeB42** [IfcOpenShell/IfcOpenShell#237]
Hello. Is it possible to federate multiple IFC files into one model using IfcOpenShell? I've asked on IRC and was instructed to ask here.

Thanks in advance.

**Comment by aothms** [comment]
Hi, answer depends on the kind of automation you expect. There is no ready made executable, but in Python/C++ there is a fair amount of code that will help you copy instances over from one file to another. Length units can be automatically converted to some extent [1]. But making sure there is a sensible project hierarchy, with `IfcProject` > `IfcBuilding` > etc. still needs to be accounted for somehow by the end-user.

Right now there is IfcConvert that can convert IFC files into various other formats, but does not emit IFC files as output. I do hope at some point there will be such a utility within the toolset of IfcOpenShell, to filter, fix, concatenate, manipulate IFC files from the command line. I am certainly open to any kind of contributions towards this.

[1] https://github.com/IfcOpenShell/IfcOpenShell/blob/master/src/ifcparse/IfcParse.cpp#L1495

**Comment by LukeB42** [comment]
@aothms thanks for your prompt reply. I'll look into federating models with IfcOpenShell manually then. Thanks for pointing that out.

One last question: Can you please update the Python wrapper for loading IFC files into IfcOpenShell via an existing file-like object instead of only having `open(fn=None)`, which expects files to to be obtained from the local filesystem?

**Comment by aothms** [comment]
> Can you please update the Python wrapper for loading IFC files into IfcOpenShell via an existing file-like object

Not sure how to accomplish that, didn't look into it. Since everything is implemented in C++ you'd somehow need to get the file handle from Python into C++. I guess it should be possible, but since Python file-like objects are such a generic concept (file, network stream, pipe, ...) there might be many intricacies in implementing this.

**Comment by LukeB42** [comment]
OK no problem and thanks.
I might be able to get a PR across containing this functionality (reading into a bytearray in Python and then buffering that within the shared library), depending on whether it's cool with my employer.
The Python file object protocol is only concerned about the existence of file operation methods and any exceptions can be expected to be handled by the user as usual. This addition would enable people to load their IFC files over the network directly into IfcOpenShell, which saves on physical disk writes.

How would you feel about merging this if I send something across to you within the week?

**Comment by aothms** [comment]
> This addition would enable people to load their IFC files over the network directly into IfcOpenShell, which saves on physical disk writes.

I agree, other ways of parsing IFC data are very welcome. Also e.g. simply from string. Saves a tempfile e.g. in this tutorial [1].

[1] http://academy.ifcopenshell.org/creating-a-simple-wall-with-property-set-and-quantity-information/

Are you planning to simply read all in Python and pass a string to C++ or are you planning on passing a stream? Both have pros and cons I'd argue.

> How would you feel about merging this if I send something across to you within the week?

Sounds very good to me

**Comment by LukeB42** [comment]
@aothms I've implemented loading from existing strings in the Python wrapper and wanted to ask the following:

1) Should this PR have a specific branch name?
2) In exchange, could you provide an example of how to federate two ifc files into one model?

Regards,
Luke

**Comment by aothms** [comment]
Cool :)

1. I don't have a preference for a branch name. I will rebase/merge it onto the master branch. I don't think the branch name will show up somewhere except possibly in a merge commit message that we can override.

2. Ok, so it turns out there were some minor issues I had to resolve, but those are out of the way (make sure you get the most recent commits). See the code below. Like I said, the bare concatenation of various files is rather trivial. But to get a sensible project hierarchy back is where things get complicated. I went as far as to remove the 2nd IfcProject, but depending on the use case, more work might be needed, e.g. merging up to building storey level and probably also merging the relationships (`IfcRelAggregates` in this case).

~~~py
import ifcopenshell

new_file = ifcopenshell.open()

old_files = list(map(ifcopenshell.open, [
    "Duplex_A_20110907_optimized.ifc",
    "Duplex_MEP_20110907_optimized.ifc"
]))

for old in old_files:
    for inst in old:
        new_file.add(inst)

projects = new_file.by_type("IfcProject")
# Process all but the first project
for p in projects[1:]:
    # Get instances that relate to this project
    refs = new_file.get_inverse(p)
    is_project = lambda inst: inst == p
    assign_first = lambda inst: projects[0]
    # Assign the first project wherever the other project is assigned
    for ref in refs:
        for index, attr in enumerate(ref):
            ref[index] = ifcopenshell.entity_instance.walk(
                is_project,
                assign_first,
                ref[index])
    # Remove from file
    new_file.remove(p)

new_file.write("merged.ifc")
~~~

**Comment by LukeB42** [comment]
@aothms Thanks for the code snippet and the tip about getting the latest commits. I've made a PR but I'm having trouble implementing a `Dup` method for copying an existing `const std::string&` instead of creating a new file descriptor -  the new prototype in `IfcFile.h` isn't being picked up during compilation or I'm not adding the prototype to the right file. What do I need to do to introduce `IfcParse::IfcFile::Dup` apart from implementing the method in `IfcParse.cpp`?

**Comment by LukeB42** [comment]
Hi @aothms. I'm sorry to be the bearer of bad news but I thought you may find this traceback useful, given the above code listing:
```
$ gdb python
GNU gdb (Ubuntu 7.11.1-0ubuntu1~16.5) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python...done.
(gdb) run federate_models.py
Starting program: /usr/bin/python federate_models.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[<ifcopenshell.file.file object at 0x7ffff2c97668>, <ifcopenshell.file.file object at 0x7ffff2c97748>]

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff37bb4d2 in ?? () from /home/luke/Development/Frontend/ifcopenshell/_ifcopenshell_wrapper.so
(gdb) bt
#0  0x00007ffff37bb4d2 in ?? () from /home/luke/Development/Frontend/ifcopenshell/_ifcopenshell_wrapper.so
#1  0x00007ffff3a50841 in ?? () from /home/luke/Development/Frontend/ifcopenshell/_ifcopenshell_wrapper.so
#2  0x00007ffff3a4b4fb in ?? () from /home/luke/Development/Frontend/ifcopenshell/_ifcopenshell_wrapper.so
#3  0x00007ffff37dc7b6 in ?? () from /home/luke/Development/Frontend/ifcopenshell/_ifcopenshell_wrapper.so
#4  0x00007ffff7976589 in _PyCFunction_FastCallDict (func_obj=0x7ffff65c3678, args=0x7ffff7e3c8d8, nargs=<optimised out>, kwargs=kwargs@entry=0x0) at Objects/methodobject.c:234
#5  0x00007ffff7976857 in _PyCFunction_FastCallKeywords (func=func@entry=0x7ffff65c3678, stack=stack@entry=0x7ffff7e3c8d8, nargs=<optimised out>, kwnames=kwnames@entry=0x0) at Objects/methodobject.c:295
#6  0x00007ffff7a09781 in call_function (pp_stack=pp_stack@entry=0x7fffffffd680, oparg=oparg@entry=2, kwnames=kwnames@entry=0x0) at Python/ceval.c:4798
#7  0x00007ffff7a0d4d7 in _PyEval_EvalFrameDefault (f=<optimised out>, throwflag=<optimised out>) at Python/ceval.c:3284
#8  0x00007ffff7a08a50 in _PyFunction_FastCall (co=<optimised out>, args=<optimised out>, nargs=2, globals=<optimised out>) at Python/ceval.c:4880
#9  0x00007ffff7a0997d in fast_function (kwnames=0x0, nargs=<optimised out>, stack=<optimised out>, func=0x7ffff2c940d0) at Python/ceval.c:4915
#10 call_function (pp_stack=pp_stack@entry=0x7fffffffd860, oparg=oparg@entry=1, kwnames=kwnames@entry=0x0) at Python/ceval.c:4819
#11 0x00007ffff7a0d4d7 in _PyEval_EvalFrameDefault (f=<optimised out>, throwflag=<optimised out>) at Python/ceval.c:3284
#12 0x00007ffff7a08a50 in _PyFunction_FastCall (co=<optimised out>, args=<optimised out>, nargs=2, globals=<optimised out>) at Python/ceval.c:4880
#13 0x00007ffff7a0997d in fast_function (kwnames=0x0, nargs=<optimised out>, stack=<optimised out>, func=0x7ffff7e46268) at Python/ceval.c:4915
#14 call_function (pp_stack=pp_stack@entry=0x7fffffffda40, oparg=oparg@entry=1, kwnames=kwnames@entry=0x0) at Python/ceval.c:4819
#15 0x00007ffff7a0d4d7 in _PyEval_EvalFrameDefault (f=<optimised out>, throwflag=<optimised out>) at Python/ceval.c:3284
#16 0x00007ffff7a093e4 in _PyEval_EvalCodeWithName (_co=_co@entry=0x7ffff7eca9c0, globals=globals@entry=0x7ffff7f4e048, locals=locals@entry=0x7ffff7f4e048, args=args@entry=0x0, argcount=argcount@entry=0, kwnames=kwnames@entry=0x0, kwargs=0x8, kwcount=0, kwstep=2, defs=0x0, defcount=0, kwdefs=0x0, closure=0x0, name=0x0, qualname=0x0) at Python/ceval.c:4128
#17 0x00007ffff7a099bf in PyEval_EvalCodeEx (_co=_co@entry=0x7ffff7eca9c0, globals=globals@entry=0x7ffff7f4e048, locals=locals@entry=0x7ffff7f4e048, args=args@entry=0x0, argcount=argcount@entry=0, kws=kws@entry=0x0, kwcount=0, defs=0x0, defcount=0, kwdefs=0x0, closure=0x0) at Python/ceval.c:4149
#18 0x00007ffff7a099eb in PyEval_EvalCode (co=co@entry=0x7ffff7eca9c0, globals=globals@entry=0x7ffff7f4e048, locals=locals@entry=0x7ffff7f4e048) at Python/ceval.c:695
#19 0x00007ffff7a3dac2 in run_mod (arena=0x7ffff7f66240, flags=0x7fffffffdd30, locals=0x7ffff7f4e048, globals=0x7ffff7f4e048, filename=0x7ffff6882a08, mod=0x63f560) at Python/pythonrun.c:980
#20 PyRun_FileExFlags (fp=fp@entry=0x68ecd0, filename_str=filename_str@entry=0x7ffff7f07868 "federate_models.py", start=start@entry=257, globals=globals@entry=0x7ffff7f4e048, locals=locals@entry=0x7ffff7f4e048, closeit=closeit@entry=1, flags=0x7fffffffdd30) at Python/pythonrun.c:933
#21 0x00007ffff7a3dc46 in PyRun_SimpleFileExFlags (fp=fp@entry=0x68ecd0, filename=<optimised out>, closeit=closeit@entry=1, flags=flags@entry=0x7fffffffdd30) at Python/pythonrun.c:396
#22 0x00007ffff7a3e083 in PyRun_AnyFileExFlags (fp=fp@entry=0x68ecd0, filename=<optimised out>, closeit=closeit@entry=1, flags=flags@entry=0x7fffffffdd30) at Python/pythonrun.c:80
#23 0x00007ffff7a59ae3 in run_file (p_cf=0x7fffffffdd30, filename=0x604280 L"federate_models.py", fp=0x68ecd0) at Modules/main.c:338
#24 Py_Main (argc=argc@entry=2, argv=argv@entry=0x603010) at Modules/main.c:809
#25 0x0000000000400ac4 in main (argc=2, argv=<optimised out>) at ./Programs/python.c:69
```

Thanks for your help, anyway.

**Comment by aothms** [comment]
The traceback is unfortunately not very useful because there are no debug symbols for `_ifcopenshell_wrapper.so` hence the `?? ()` for function names. Would it be possible to send over a complete package of script and input files so that I can reproduce this locally?

**Comment by Moult** [comment]
Bumping this issue that I've used this recipe which has worked on quite a few large files produced from Revit:

https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcpatch/recipes/MergeProject.py

**Comment by gschleusner** [comment]
> Bumping this issue that I've used this recipe which has worked on quite a few large files produced from Revit:

@Moult is there a public link to this MergeProject example?

**Comment by aothms** [comment]
@gschleusner see here https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcpatch/ifcpatch/recipes/MergeProject.py

There's quite a few approaches scattered over the code base there is also https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/project/append_asset.py#L90

The fundamental operation of copying definitions from one file to the other is well supported in IfcOpenShell, but the exact functionality on how to handle the merging and what relationships to traverse still hasn't made it into one cohesive approach.

**Comment by Moult** [comment]
@gschleusner Yeah the simple fact is that nobody can agree on what it means to merge a project. Merge by guid? By spatial element? By type?

The MergeProject provides an reliable way of merging two projects - but it's very "hands-off". It merges the project, and every thing else is simply appended - no combination of elements.

There have been discussions on providing options for known usecases: e.g. by guid, or merging portions. The append_asset usecase is a good example of selective merging.

In practice I've found that there are a huge number of problems to solve before merging. A particularly notorious bug is where Revit duplicates the types. In ArchiCAD you can't control your project GlobalId. In Tekla the out of the box spatial structure makes life a pain for them to integrate with other trades.

In short - I always end up writing a bespoke script. IfcOpenShell comes with all the batteries you need to make writing such a script really quite pleasant and powerful, but it is still bespoke on a per-project basis.

We'll probably stay that way for a while until the industry catches up to ISO standards.

**Comment by Moult** [comment]
This can probably be closed now as there are a variety of utility functions to write your own or just use the MergeProject recipe.
